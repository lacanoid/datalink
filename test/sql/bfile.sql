\pset null _null_

SET client_min_messages = notice;
SET search_path = public,datalink;

select getlength(dlvalue('/var/www/datalink/test2.txt'));

select substr(dlvalue('/var/www/datalink/test2.txt')),
       substr(dlvalue('/var/www/datalink/test2.txt'),12),
       substr(dlvalue('/var/www/datalink/test2.txt'),12,8)
       ;

create domain bfile datalink(2);
create table bfiles ( bfile bfile );

insert into bfiles (bfile) values (dlvalue('test1.txt','www'));
insert into bfiles (bfile) values (dlvalue('test2.txt','www'));
insert into bfiles (bfile) values (dlvalue('test3.txt#11111111-2222-3333-4444-abecedabeced','www'));

create domain bfile2 bfile;
create table bfiles2 ( bfile bfile2 );

insert into bfiles2 (bfile) values (dlvalue('test1.txt','www'));
insert into bfiles2 (bfile) values (dlvalue('test4.txt','www'));

select * from datalink.types;

select regclass,column_name,lco,regtype from datalink.dl_columns where regtype <> 'datalink'::regtype order by 1,2;

select substr(dlvalue('test3.txt#11111111-2222-3333-4444-abecedabeced','www'),1,19);

select getlength(bfile),substr(bfile,1,5),filepath(bfile),fileexists(bfile),filegetname(bfile)
  from bfiles;

select read_text(dlvalue('test3.txt#11111111-2222-3333-4444-abecedabeced','www'),1,68);
select read_text(dlvalue('test3.txt#11111111-2222-3333-4444-abecedabeced','www'),48,48);

drop table bfiles;
truncate bfiles2;

-- test text file reads

select read_text('/var/www/datalink/test2.txt');
select read_text('/var/www/datalink/test2.txt',11);
select read_text('/var/www/datalink/test2.txt',11,10);
select read_text(dlvalue('/var/www/datalink/test2.txt'));
select read_text(dlvalue('/var/www/datalink/test2.txt'),11);
select read_text(dlvalue('/var/www/datalink/test2.txt'),11,10);
select read_text(dlvalue('https://raw.githubusercontent.com/lacanoid/datalink/refs/heads/master/docs/utf8.txt'));
select read_text(dlvalue('https://raw.githubusercontent.com/lacanoid/datalink/refs/heads/master/docs/utf8.txt'),11);
select read_text(dlvalue('https://raw.githubusercontent.com/lacanoid/datalink/refs/heads/master/docs/utf8.txt'),11,10);

select * from read_lines('/var/www/datalink/test2.txt');

-- test binary file reads

select read('/var/www/datalink/test2.txt');
select read('/var/www/datalink/test2.txt',4);
select read('/var/www/datalink/test2.txt',4,10);
select read(dlvalue('/var/www/datalink/test2.txt'));
select read(dlvalue('/var/www/datalink/test2.txt'),4);
select read(dlvalue('/var/www/datalink/test2.txt'),4,10);
select read(dlvalue('https://raw.githubusercontent.com/lacanoid/datalink/refs/heads/master/docs/utf8.txt'));
select read(dlvalue('https://raw.githubusercontent.com/lacanoid/datalink/refs/heads/master/docs/utf8.txt'),4);
select read(dlvalue('https://raw.githubusercontent.com/lacanoid/datalink/refs/heads/master/docs/utf8.txt'),4,10);

-- test checking for updated files

insert into my_table2 values (dlvalue('/var/www/datalink/test2.txt'));
select datalink.has_updated(link) from my_table2;
\! touch /var/www/datalink/test2.txt
select datalink.has_updated(link) from my_table2;
select datalink.has_updated(link) from my_table2;
select datalink.has_updated(dlvalue('/var/www/datalink/test2.txt'));
update my_table2 set link=dlnewcopy(link);
select datalink.has_updated(link) from my_table2;
select datalink.has_updated(dlvalue('/var/www/datalink/test2.txt'));
truncate my_table2;
select datalink.has_updated(dlvalue('/var/www/datalink/test2.txt'));

select datalink.has_updated('/etc/issue');

-- test link construction types

create table files ( link datalink(2) );
insert into files values (dlvalue('/var/www/datalink/test1.txt'));
update files  set link = dlnewcopy(link);
insert into files values (dlvalue('/var/www/datalink/test2.txt'));
insert into files values (dlreplacecontent('/var/www/datalink/test6.txt','http://www.github.com/robots.txt'));
insert into files values (datalink.write_text(dlvalue('test5.txt','www'),'This is a test file 5'));

select cons,path from datalink.dl_linked_files order by path;

drop table files;

-- test curl
select body from datalink.curl_get('https://raw.githubusercontent.com/lacanoid/datalink/refs/heads/master/docs/utf8.txt');
select content_type,size from datalink.curl_save(
  '/var/www/datalink/installcheck_curl_save.txt',
  'https://raw.githubusercontent.com/lacanoid/datalink/refs/heads/master/docs/utf8.txt',1);
select content_type,size from datalink.curl_save(
  '/var/www/datalink/installcheck_curl_save.txt',
  'https://raw.githubusercontent.com/lacanoid/datalink/refs/heads/master/docs/utf8.txt',1); -- error
select * from datalink.read_text('/var/www/datalink/installcheck_curl_save.txt');

-- test transactional writes and reads

create table l1 (link datalink(2));
begin;
insert into l1 values (datalink.write_text(dlvalue('installcheck_xact1.1.txt','www'),'This is a write test file'));
select substr(link) from l1;
update l1 set link = datalink.write_text(link,'This is a write test file 1');
select substr(link) from l1;
abort;
begin;
insert into l1 values (dlvalue(datalink.write_text('/var/www/datalink/installcheck_xact1.2.txt','This is a write test file')));
select substr(link) from l1;
update l1 set link = datalink.write_text(link,'This is a write test file 2');
select substr(link) from l1;
abort;
begin;
select substr(dlvalue('https://raw.githubusercontent.com/lacanoid/datalink/refs/heads/master/docs/utf8.txt'));
insert into l1 values (dlreplacecontent(
  '/var/www/datalink/installcheck_xact1.3.txt',
  'https://raw.githubusercontent.com/lacanoid/datalink/refs/heads/master/docs/utf8.txt'));
select substr(link) from l1;
update l1 set link = datalink.write_text(link,'This is a write test file 3');
select substr(link) from l1;
abort;
drop table l1;
