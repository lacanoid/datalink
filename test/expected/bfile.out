\pset null _null_
SET client_min_messages = notice;
SET search_path = public,datalink;
select getlength(dlvalue('/var/www/datalink/test2.txt'));
 getlength 
-----------
       177
(1 row)

select substr(dlvalue('/var/www/datalink/test2.txt')),
       substr(dlvalue('/var/www/datalink/test2.txt'),12),
       substr(dlvalue('/var/www/datalink/test2.txt'),12,8)
       ;
                                    substr                                    |                              substr                               |  substr  
------------------------------------------------------------------------------+-------------------------------------------------------------------+----------
 Kulturnik je portal, namenjen lažji in orientaciji po kulturi v slovenščini.+| e portal, namenjen lažji in orientaciji po kulturi v slovenščini.+| e portal
 Some 16bit emojis: ⛔⛳                                                     +| Some 16bit emojis: ⛔⛳                                          +| 
 Some 32bit emojis: 🪙🩲                                                     +| Some 32bit emojis: 🪙🩲                                          +| 
 Some chinese: 𠀀𠀁𠀂𠀃                                                      +| Some chinese: 𠀀𠀁𠀂𠀃                                           +| 
 The end.                                                                    +| The end.                                                         +| 
                                                                             +|                                                                  +| 
                                                                             +|                                                                  +| 
                                                                             +|                                                                  +| 
                                                                              |                                                                   | 
(1 row)

create domain bfile datalink(2);
create table bfiles ( bfile bfile );
NOTICE:  DATALINK DDL:TRIGGER on bfiles
insert into bfiles (bfile) values (dlvalue('test1.txt','www'));
NOTICE:  DATALINK LINK:/var/www/datalink/test1.txt
insert into bfiles (bfile) values (dlvalue('test2.txt','www'));
NOTICE:  DATALINK LINK:/var/www/datalink/test2.txt
insert into bfiles (bfile) values (dlvalue('test3.txt#11111111-2222-3333-4444-abecedabeced','www'));
NOTICE:  DATALINK LINK:/var/www/datalink/test3.txt
create domain bfile2 bfile;
create table bfiles2 ( bfile bfile2 );
NOTICE:  DATALINK DDL:TRIGGER on bfiles2
insert into bfiles2 (bfile) values (dlvalue('test1.txt','www'));
ERROR:  DATALINK EXCEPTION - external file already linked
DETAIL:  from bfiles.bfile as '/var/www/datalink/test1.txt'
CONTEXT:  PL/pgSQL function dl_file_link(file_path,dl_token,"char",dl_lco,regclass,name) line 81 at RAISE
SQL statement "SELECT datalink.dl_file_link(my_path,my_token,cons,link_options,regclass,column_name)"
PL/pgSQL function dl_datalink_ref(datalink,dl_lco,regclass,name) line 92 at PERFORM
PL/pgSQL function dl_trigger_table() line 94 at assignment
insert into bfiles2 (bfile) values (dlvalue('test4.txt','www'));
NOTICE:  DATALINK LINK:/var/www/datalink/test4.txt
select * from datalink.types;
 regtype  | typtype | typmod | depth 
----------+---------+--------+-------
 datalink | b       | _null_ |     0
 file     | d       | _null_ |     1
 rfile    | d       |      1 |     1
 efile    | d       |      2 |     1
 bfile    | d       |      2 |     1
 bfile2   | d       |      2 |     2
(6 rows)

select regclass,column_name,lco,regtype from datalink.dl_columns where regtype <> 'datalink'::regtype order by 1,2;
 regclass | column_name | lco | regtype 
----------+-------------+-----+---------
 efiles0  | file        |   0 | file
 efiles1  | file        |   1 | rfile
 efiles2  | file        |   2 | efile
 bfiles   | bfile       |   2 | bfile
 bfiles2  | bfile       |   2 | bfile2
(5 rows)

select substr(dlvalue('test3.txt#11111111-2222-3333-4444-abecedabeced','www'),1,19);
       substr        
---------------------
 © 2018 Žiga Kranjec
(1 row)

select getlength(bfile),substr(bfile,1,5),filepath(bfile),fileexists(bfile),filegetname(bfile)
  from bfiles;
 getlength | substr |                             filepath                             | fileexists |            filegetname             
-----------+--------+------------------------------------------------------------------+------------+------------------------------------
         6 | Hello  | /var/www/datalink/test1.txt                                      |          1 | (www,test1.txt,/var/www/datalink/)
       177 | Kultu  | /var/www/datalink/test2.txt                                      |          1 | (www,test2.txt,/var/www/datalink/)
       939 | © 201  | /var/www/datalink/test3.txt#11111111-2222-3333-4444-abecedabeced |          1 | (www,test3.txt,/var/www/datalink/)
(3 rows)

select read_text(dlvalue('test3.txt#11111111-2222-3333-4444-abecedabeced','www'),1,68);
                  read_text                  
---------------------------------------------
 © 2018 Žiga Kranjec <lacanoid@ljudmila.org>+
                                            +
 Permission to use, co
(1 row)

select read_text(dlvalue('test3.txt#11111111-2222-3333-4444-abecedabeced','www'),48,48);
                    read_text                     
--------------------------------------------------
 Permission to use, copy, modify, and distribute 
(1 row)

drop table bfiles;
NOTICE:  DATALINK UNLINK:/var/www/datalink/test1.txt
NOTICE:  DATALINK UNLINK:/var/www/datalink/test2.txt
NOTICE:  DATALINK UNLINK:/var/www/datalink/test3.txt
truncate bfiles2;
NOTICE:  DATALINK UNLINK:/var/www/datalink/test4.txt
-- test text file reads
select read_text('/var/www/datalink/test2.txt');
                                  read_text                                   
------------------------------------------------------------------------------
 Kulturnik je portal, namenjen lažji in orientaciji po kulturi v slovenščini.+
 Some 16bit emojis: ⛔⛳                                                     +
 Some 32bit emojis: 🪙🩲                                                     +
 Some chinese: 𠀀𠀁𠀂𠀃                                                      +
 The end.                                                                    +
                                                                             +
                                                                             +
                                                                             +
 
(1 row)

select read_text('/var/www/datalink/test2.txt',11);
                             read_text                              
--------------------------------------------------------------------
 je portal, namenjen lažji in orientaciji po kulturi v slovenščini.+
 Some 16bit emojis: ⛔⛳                                           +
 Some 32bit emojis: 🪙🩲                                           +
 Some chinese: 𠀀𠀁𠀂𠀃                                            +
 The end.                                                          +
                                                                   +
                                                                   +
                                                                   +
 
(1 row)

select read_text('/var/www/datalink/test2.txt',11,10);
 read_text  
------------
 je portal,
(1 row)

select read_text(dlvalue('/var/www/datalink/test2.txt'));
                                  read_text                                   
------------------------------------------------------------------------------
 Kulturnik je portal, namenjen lažji in orientaciji po kulturi v slovenščini.+
 Some 16bit emojis: ⛔⛳                                                     +
 Some 32bit emojis: 🪙🩲                                                     +
 Some chinese: 𠀀𠀁𠀂𠀃                                                      +
 The end.                                                                    +
                                                                             +
                                                                             +
                                                                             +
 
(1 row)

select read_text(dlvalue('/var/www/datalink/test2.txt'),11);
                             read_text                              
--------------------------------------------------------------------
 je portal, namenjen lažji in orientaciji po kulturi v slovenščini.+
 Some 16bit emojis: ⛔⛳                                           +
 Some 32bit emojis: 🪙🩲                                           +
 Some chinese: 𠀀𠀁𠀂𠀃                                            +
 The end.                                                          +
                                                                   +
                                                                   +
                                                                   +
 
(1 row)

select read_text(dlvalue('/var/www/datalink/test2.txt'),11,10);
 read_text  
------------
 je portal,
(1 row)

select read_text(dlvalue('https://raw.githubusercontent.com/lacanoid/datalink/refs/heads/master/docs/utf8.txt'));
                                  read_text                                   
------------------------------------------------------------------------------
 Kulturnik je portal, namenjen lažji in orientaciji po kulturi v slovenščini.+
 Some 16bit emojis: ⛔⛳                                                     +
 Some 32bit emojis: 🪙🩲                                                     +
 Some chinese: 𠀀𠀁𠀂𠀃                                                      +
 The end.                                                                    +
                                                                             +
                                                                             +
                                                                             +
 
(1 row)

select read_text(dlvalue('https://raw.githubusercontent.com/lacanoid/datalink/refs/heads/master/docs/utf8.txt'),11);
                             read_text                              
--------------------------------------------------------------------
 je portal, namenjen lažji in orientaciji po kulturi v slovenščini.+
 Some 16bit emojis: ⛔⛳                                           +
 Some 32bit emojis: 🪙🩲                                           +
 Some chinese: 𠀀𠀁𠀂𠀃                                            +
 The end.                                                          +
                                                                   +
                                                                   +
                                                                   +
 
(1 row)

select read_text(dlvalue('https://raw.githubusercontent.com/lacanoid/datalink/refs/heads/master/docs/utf8.txt'),11,10);
 read_text  
------------
 je portal,
(1 row)

select * from read_lines('/var/www/datalink/test2.txt');
 i |  o  |                                     line                                     
---+-----+------------------------------------------------------------------------------
 1 |   1 | Kulturnik je portal, namenjen lažji in orientaciji po kulturi v slovenščini.
 2 |  78 | Some 16bit emojis: ⛔⛳
 3 | 100 | Some 32bit emojis: 🪙🩲
 4 | 122 | Some chinese: 𠀀𠀁𠀂𠀃
 5 | 141 | The end.
 6 | 150 | 
 7 | 151 | 
 8 | 152 | 
(8 rows)

-- test binary file reads
select read('/var/www/datalink/test2.txt');
                                                                                                                                                                                 read                                                                                                                                                                                 
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 \x4b756c7475726e696b206a6520706f7274616c2c206e616d656e6a656e206c61c5be6a6920696e206f7269656e746163696a6920706f206b756c74757269207620736c6f76656ec5a1c48d696e692e0a536f6d6520313662697420656d6f6a69733a20e29b94e29bb30a536f6d6520333262697420656d6f6a69733a20f09faa99f09fa9b20a536f6d65206368696e6573653a20f0a08080f0a08081f0a08082f0a080830a54686520656e642e0a0a0a0a
(1 row)

select read('/var/www/datalink/test2.txt',4);
                                                                                                                                                                              read                                                                                                                                                                              
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 \x7475726e696b206a6520706f7274616c2c206e616d656e6a656e206c61c5be6a6920696e206f7269656e746163696a6920706f206b756c74757269207620736c6f76656ec5a1c48d696e692e0a536f6d6520313662697420656d6f6a69733a20e29b94e29bb30a536f6d6520333262697420656d6f6a69733a20f09faa99f09fa9b20a536f6d65206368696e6573653a20f0a08080f0a08081f0a08082f0a080830a54686520656e642e0a0a0a0a
(1 row)

select read('/var/www/datalink/test2.txt',4,10);
          read          
------------------------
 \x7475726e696b206a6520
(1 row)

select read(dlvalue('/var/www/datalink/test2.txt'));
                                                                                                                                                                                 read                                                                                                                                                                                 
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 \x4b756c7475726e696b206a6520706f7274616c2c206e616d656e6a656e206c61c5be6a6920696e206f7269656e746163696a6920706f206b756c74757269207620736c6f76656ec5a1c48d696e692e0a536f6d6520313662697420656d6f6a69733a20e29b94e29bb30a536f6d6520333262697420656d6f6a69733a20f09faa99f09fa9b20a536f6d65206368696e6573653a20f0a08080f0a08081f0a08082f0a080830a54686520656e642e0a0a0a0a
(1 row)

select read(dlvalue('/var/www/datalink/test2.txt'),4);
                                                                                                                                                                              read                                                                                                                                                                              
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 \x7475726e696b206a6520706f7274616c2c206e616d656e6a656e206c61c5be6a6920696e206f7269656e746163696a6920706f206b756c74757269207620736c6f76656ec5a1c48d696e692e0a536f6d6520313662697420656d6f6a69733a20e29b94e29bb30a536f6d6520333262697420656d6f6a69733a20f09faa99f09fa9b20a536f6d65206368696e6573653a20f0a08080f0a08081f0a08082f0a080830a54686520656e642e0a0a0a0a
(1 row)

select read(dlvalue('/var/www/datalink/test2.txt'),4,10);
          read          
------------------------
 \x7475726e696b206a6520
(1 row)

select read(dlvalue('https://raw.githubusercontent.com/lacanoid/datalink/refs/heads/master/docs/utf8.txt'));
                                                                                                                                                                                 read                                                                                                                                                                                 
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 \x4b756c7475726e696b206a6520706f7274616c2c206e616d656e6a656e206c61c5be6a6920696e206f7269656e746163696a6920706f206b756c74757269207620736c6f76656ec5a1c48d696e692e0a536f6d6520313662697420656d6f6a69733a20e29b94e29bb30a536f6d6520333262697420656d6f6a69733a20f09faa99f09fa9b20a536f6d65206368696e6573653a20f0a08080f0a08081f0a08082f0a080830a54686520656e642e0a0a0a0a
(1 row)

select read(dlvalue('https://raw.githubusercontent.com/lacanoid/datalink/refs/heads/master/docs/utf8.txt'),4);
                                                                                                                                                                              read                                                                                                                                                                              
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 \x7475726e696b206a6520706f7274616c2c206e616d656e6a656e206c61c5be6a6920696e206f7269656e746163696a6920706f206b756c74757269207620736c6f76656ec5a1c48d696e692e0a536f6d6520313662697420656d6f6a69733a20e29b94e29bb30a536f6d6520333262697420656d6f6a69733a20f09faa99f09fa9b20a536f6d65206368696e6573653a20f0a08080f0a08081f0a08082f0a080830a54686520656e642e0a0a0a0a
(1 row)

select read(dlvalue('https://raw.githubusercontent.com/lacanoid/datalink/refs/heads/master/docs/utf8.txt'),4,10);
          read          
------------------------
 \x7475726e696b206a6520
(1 row)

-- test checking for updated files
insert into my_table2 values (dlvalue('/var/www/datalink/test2.txt'));
NOTICE:  DATALINK LINK:/var/www/datalink/test2.txt
select datalink.has_updated(link) from my_table2;
 has_updated 
-------------
           0
(1 row)

\! touch /var/www/datalink/test2.txt
select datalink.has_updated(link) from my_table2;
 has_updated 
-------------
           1
(1 row)

select datalink.has_updated(link) from my_table2;
 has_updated 
-------------
           1
(1 row)

select datalink.has_updated(dlvalue('/var/www/datalink/test2.txt'));
 has_updated 
-------------
           1
(1 row)

update my_table2 set link=dlnewcopy(link);
NOTICE:  DATALINK UNLINK:/var/www/datalink/test2.txt
NOTICE:  DATALINK LINK:/var/www/datalink/test2.txt
select datalink.has_updated(link) from my_table2;
 has_updated 
-------------
           0
(1 row)

select datalink.has_updated(dlvalue('/var/www/datalink/test2.txt'));
 has_updated 
-------------
           0
(1 row)

truncate my_table2;
NOTICE:  DATALINK UNLINK:/var/www/datalink/test2.txt
select datalink.has_updated(dlvalue('/var/www/datalink/test2.txt'));
WARNING:  DATALINK WARNING - external file not linked
DETAIL:  function has_updated() only works with linked files
HINT:  store a datalink referencing this file in a column with INTEGRITY ALL option
 has_updated 
-------------
      _null_
(1 row)

select datalink.has_updated('/etc/issue');
WARNING:  DATALINK WARNING - external file not linked
DETAIL:  function has_updated() only works with linked files
HINT:  store a datalink referencing this file in a column with INTEGRITY ALL option
 has_updated 
-------------
      _null_
(1 row)

-- test link construction types
create table files ( link datalink(2) );
NOTICE:  DATALINK DDL:TRIGGER on files
insert into files values (dlvalue('/var/www/datalink/test1.txt'));
NOTICE:  DATALINK LINK:/var/www/datalink/test1.txt
update files  set link = dlnewcopy(link);
NOTICE:  DATALINK UNLINK:/var/www/datalink/test1.txt
NOTICE:  DATALINK LINK:/var/www/datalink/test1.txt
insert into files values (dlvalue('/var/www/datalink/test2.txt'));
NOTICE:  DATALINK LINK:/var/www/datalink/test2.txt
insert into files values (dlreplacecontent('/var/www/datalink/test6.txt','http://www.github.com/robots.txt'));
NOTICE:  DATALINK LINK:/var/www/datalink/test6.txt
insert into files values (datalink.write_text(dlvalue('test5.txt','www'),'This is a test file 5'));
NOTICE:  DATALINK LINK:/var/www/datalink/test5.txt
select cons,path from datalink.dl_linked_files order by path;
  cons  |            path             
--------+-----------------------------
 n      | /var/www/datalink/test1.txt
 _null_ | /var/www/datalink/test2.txt
 w      | /var/www/datalink/test5.txt
 r      | /var/www/datalink/test6.txt
(4 rows)

drop table files;
NOTICE:  DATALINK UNLINK:/var/www/datalink/test1.txt
NOTICE:  DATALINK UNLINK:/var/www/datalink/test2.txt
NOTICE:  DATALINK UNLINK:/var/www/datalink/test6.txt
NOTICE:  DATALINK UNLINK:/var/www/datalink/test5.txt
-- test curl
select body from datalink.curl_get('https://raw.githubusercontent.com/lacanoid/datalink/refs/heads/master/docs/utf8.txt');
                                     body                                     
------------------------------------------------------------------------------
 Kulturnik je portal, namenjen lažji in orientaciji po kulturi v slovenščini.+
 Some 16bit emojis: ⛔⛳                                                     +
 Some 32bit emojis: 🪙🩲                                                     +
 Some chinese: 𠀀𠀁𠀂𠀃                                                      +
 The end.                                                                    +
                                                                             +
                                                                             +
                                                                             +
 
(1 row)

select content_type,size from datalink.curl_save(
  '/var/www/datalink/installcheck/curl_save.txt',
  'https://raw.githubusercontent.com/lacanoid/datalink/refs/heads/master/docs/utf8.txt',1);
       content_type        | size 
---------------------------+------
 text/plain; charset=utf-8 |  177
(1 row)

select content_type,size from datalink.curl_save(
  '/var/www/datalink/installcheck/curl_save.txt',
  'https://raw.githubusercontent.com/lacanoid/datalink/refs/heads/master/docs/utf8.txt',1); -- error
ERROR:  DATALINK EXCEPTIION - file exists
FILE:  /var/www/datalink/installcheck/curl_save.txt
CONTEXT:  PL/Perl function "curl_save"
select * from datalink.read_text('/var/www/datalink/installcheck/curl_save.txt');
                                  read_text                                   
------------------------------------------------------------------------------
 Kulturnik je portal, namenjen lažji in orientaciji po kulturi v slovenščini.+
 Some 16bit emojis: ⛔⛳                                                     +
 Some 32bit emojis: 🪙🩲                                                     +
 Some chinese: 𠀀𠀁𠀂𠀃                                                      +
 The end.                                                                    +
                                                                             +
                                                                             +
                                                                             +
 
(1 row)

-- test transactional writes and reads
create table l1 (link datalink(2));
NOTICE:  DATALINK DDL:TRIGGER on l1
begin;
insert into l1 values (datalink.write_text(dlvalue('installcheck/xact1.1.txt','www'),'This is a write test file'));
NOTICE:  DATALINK LINK:/var/www/datalink/installcheck/xact1.1.txt
select substr(link) from l1;
          substr           
---------------------------
 This is a write test file
(1 row)

update l1 set link = datalink.write_text(link,'This is a write test file 1');
NOTICE:  DATALINK UNLINK:/var/www/datalink/installcheck/xact1.1.txt
NOTICE:  DATALINK LINK:/var/www/datalink/installcheck/xact1.1.txt
select substr(link) from l1;
           substr            
-----------------------------
 This is a write test file 1
(1 row)

abort;
begin;
insert into l1 values (dlvalue(datalink.write_text('/var/www/datalink/installcheck/xact1.2.txt','This is a write test file')));
NOTICE:  DATALINK LINK:/var/www/datalink/installcheck/xact1.2.txt
select substr(link) from l1;
          substr           
---------------------------
 This is a write test file
(1 row)

update l1 set link = datalink.write_text(link,'This is a write test file 2');
NOTICE:  DATALINK UNLINK:/var/www/datalink/installcheck/xact1.2.txt
NOTICE:  DATALINK LINK:/var/www/datalink/installcheck/xact1.2.txt
select substr(link) from l1;
           substr            
-----------------------------
 This is a write test file 2
(1 row)

abort;
begin;
select substr(dlvalue('https://raw.githubusercontent.com/lacanoid/datalink/refs/heads/master/docs/utf8.txt'));
                                    substr                                    
------------------------------------------------------------------------------
 Kulturnik je portal, namenjen lažji in orientaciji po kulturi v slovenščini.+
 Some 16bit emojis: ⛔⛳                                                     +
 Some 32bit emojis: 🪙🩲                                                     +
 Some chinese: 𠀀𠀁𠀂𠀃                                                      +
 The end.                                                                    +
                                                                             +
                                                                             +
                                                                             +
 
(1 row)

insert into l1 values (dlreplacecontent(
  '/var/www/datalink/installcheck/xact1.3.txt',
  'https://raw.githubusercontent.com/lacanoid/datalink/refs/heads/master/docs/utf8.txt'));
NOTICE:  DATALINK LINK:/var/www/datalink/installcheck/xact1.3.txt
select substr(link) from l1;
                                    substr                                    
------------------------------------------------------------------------------
 Kulturnik je portal, namenjen lažji in orientaciji po kulturi v slovenščini.+
 Some 16bit emojis: ⛔⛳                                                     +
 Some 32bit emojis: 🪙🩲                                                     +
 Some chinese: 𠀀𠀁𠀂𠀃                                                      +
 The end.                                                                    +
                                                                             +
                                                                             +
                                                                             +
 
(1 row)

update l1 set link = datalink.write_text(link,'This is a write test file 3');
NOTICE:  DATALINK UNLINK:/var/www/datalink/installcheck/xact1.3.txt
NOTICE:  DATALINK LINK:/var/www/datalink/installcheck/xact1.3.txt
select substr(link) from l1;
           substr            
-----------------------------
 This is a write test file 3
(1 row)

abort;
drop table l1;
